### PR TITLE
Small fixes to make CI more robust

### PR DESCRIPTION
**Pull request by lhstrh**
- [x] Determine when to report coverage based on `github.repository`
- [x] Replace the wretry action (which appears to be unreliable), with a custom retry mechanism based on multiple steps.

### CODE REVIEW

**Review by cmnrd — CHANGES_REQUESTED** [review]
Did you investigate the cause of the error in https://github.com/lf-lang/lingua-franca/pull/1894?

I don't think we should ignore codecov errors. If we do this, we don't notice if there are any problems in the reporting and whether we can trust the coverage report. Also see the discussion in https://github.com/lf-lang/lingua-franca/pull/1868.

I also think we should report on all platforms, not just Linux. We have some platform dependent code in our code base and will only get full coverage if we report on all platforms. https://github.com/lf-lang/lingua-franca/pull/1868 deliberately removed the platform check.

**Comment by lhstrh** [comment]
Shall we then replace `if: ${{ github.repository == 'lf-lang/lingua-franca' && matrix.platform == 'ubuntu-latest' }}` with `if: ${{ github.repository == 'lf-lang/lingua-franca' }}` and merge this?

**Comment by lhstrh** [comment]
Note this failure: https://github.com/lf-lang/lingua-franca/actions/runs/5512578023/jobs/10049603441?pr=1895

I think this will become a major annoyance if we don't ignore failures, tbh.

**Comment by cmnrd** [comment]
This failed, because this particular report (on reactor-cpp) only works on Linux. Here we need the check on the platform (Ubuntu), and this check was there before the chanhges in this PR.

File: .github/workflows/cpp-tests.yml [hunk 1]
```diff
@@ -70,4 +70,4 @@ jobs:
         uses: ./.github/actions/report-code-coverage
         with:
           files: reactor-cpp.info
-        if: ${{ !inputs.compiler-ref && matrix.platform == 'ubuntu-latest' }}  # i.e., if this is part of the main repo's CI
+        if: ${{ github.repository == 'lf-lang/lingua-franca' }}
```
Comment (lhstrh) [inline .github/workflows/cpp-tests.yml:73]:
```suggestion
        if: ${{ github.repository == 'lf-lang/lingua-franca' && matrix.platform == 'ubuntu-latest' }}
```

**Review by cmnrd — APPROVED** [review]

**Pull request merged**